hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockDataStreamOutput.java

```diff
@@ -259,6 +268,19 @@ public void write(ByteBuffer b, int off, int len) throws IOException {
     }
 
     final StreamBuffer buf = new StreamBuffer(b, off, len);
+
+    // Controls the number of buffers on the client.
+    if (futures.size() >= streamWindowSize) {
```

Comment:
Currently, the buffer is removed only after it is committed.  After this change, the buffer can be removed after it is acked but not yet committed.  If some buffers are removed before commit, the retry won't work.